x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/CrossClusterAccessAuthenticationService.java

```diff
@@ -108,6 +111,71 @@ public void authenticate(final String action, final TransportRequest request, fi
         }
     }
 
+    public void tryAuthenticateCredentialsHeader(Map<String, String> headers, ActionListener<Void> listener) {
+        doTryAuthenticateCredentialsHeader(clusterService.threadPool().getThreadContext(), headers, listener);
+    }
+
+    // visibility for testing
+    public void doTryAuthenticateCredentialsHeader(
+        ThreadContext threadContext,
+        Map<String, String> headers,
+        ActionListener<Void> listener
+    ) {
+        final ApiKeyService.ApiKeyCredentials apiKeyCredentials;
+        try {
+            apiKeyCredentials = getApiKeyCredentialsFromHeaders(headers);
+        } catch (Exception ex) {
+            listener.onFailure(Exceptions.authenticationError("failed to parse cross cluster credentials header", ex));
+            return;
+        }
```

Comment:
I think we might want to throw exception here instead of calling the listener. Using the listener implies async handling which is less efficient and not really necessary when we can just terminate either the request or the connection in the same thread.

Maybe we could split the token extraction into its own method and let the caller to orchestrate token extraction and authentication. It is also more similar to how `Authenticator` works.

Reply:
I've split these out as you suggested. Original intention was to keep everything consolidated but having them separate is good too and makes tests easier. 

I think having these separate is preferable to a method that takes a failure handler but can also throw imo.